### DISCUSSION TITLE
Any idea on what 3D_final shaders are?

### DISCUSSION THREAD

**Discussion by Trainmaster9977** [ScanMountGoat/Smush-Material-Research#42] [Q&A] [answered]
Hello-

In my efforts to jump through hoops, I realized that the 3d_final shaders exist and might hold the key to some pipeline/blending stuff that the nurpdb and the nufxlb....maybe. Not really sure honestly but want to look into it. But that brings up the question of

Do any of us know what they even are?

I don't think I've seen them outside of the shader dump, besides maybe some 3d_final stuff in the nurpdb, though even that I don't think had the numbers. Even if I did my analyzation and simplification of the code, it wouldn't help much if I don't know what they even are or where they're used.

My best guess is that maybe they correspond to the other material shaders we're familiar with, and are like...groups maybe? the numbers in their filenames maybe being the first digits of the material shaders? I'm not super confident in that though, and its a guess at best.

So yeah, are these alien to you too, or do you have any info on em?

**Comment by ScanMountGoat** [comment, marked as answer]
What are you trying to do that can't be done just by editing one of the SFX_PBS shaders? The blending portion of the pipeline state is controlled via the materials. Are you sure that this effect requires multiple render passes?

Smash Ultimate uses a basic forward renderer, so all the lighting and shading is done by a particular material's shader. Some games use "deferred rendering" where most of the shading is "deferred" to a pass that uses screen textures for shading inputs. I haven't looked into the other nushd files much since all the "interesting" code is done with SFX_PBS shaders. I also haven't located the shader code for bloom or post processing, so you may find that in the other nushdb files. If you do manage to find code for a "global" shader, editing it may have unintended consequences on the rest of the game's visuals. 

The shader names are just strings and shouldn't be parsed in any way. We can safely extract the render pass from SFX_PBS shaders because variants of each shader name exist in the nufxlb for each of the four render passes. There may be patterns in the name generation for 3d_final, but I wouldn't rely on them. I also wouldn't expect all of these shaders to actually be used in game. You may also find shaders not related to models like the shaders used for UI rendering.

**Reply by Trainmaster9977** [reply]
As for what I'm trying to do, it's to jump through some hoops currently- my rewritten shader seemed to be done until I realized that bloom exists, so I'm trying to find a way to somehow store the effects and then layer them onto a one particular material that's done with the near pass. I might be able to finagle a way to make it work if i can fully understand the actual pipeline and how to change it. 

And yeah, while i have found the code for the bloom pass itself, that's only what's done with it once the main screen texture gets passed to it- how it's passed to it is a bit more unclear, given everything uses textures 1-16 and understanding it to the point of rewriting the pipeline a bit is proving elusive.

Anyways, as for the blending itself, that isn't done in the code for each material as far as I can tell. The main screen texture isn't passed to the shader code in renderdoc, and the layering happens in a separate step in renderdoc's pipeline than the pixel shader. Additionally, while there seems to be additional bytes that kind of look like code past the end of the main code in the shaders, I have documented one of the sfx_pbs shaders (and said documentation was supposed to be released after I finished up my cell shader mod, but yknow, the effects mean I haven't finished it) and the code there doesn't seem to have anything to do with blending- it's output is probably just passed to the actual thing that does the blending. 

At the end of the day, while I know stuff can be grouped up and passed to a shader- given the different render passes are stored and then passed to the next render pass, trying to actually figure out how it all works in terms of the pipeline is a bit difficult lol. the nurpdb probably has the answers but I'm saving it for last because every time I take a look at that file, it scares me

If it means anything though, in the nufxlb, there is a "final" render pass, listed before the rest of the near far opaque etc variants, and this applies to each material. Iirc I tried setting a material to use that in game, but it was just transparent. So not sure what's going on with that but maybe that's what the 3d_final is referring to? Idk.

**Reply by ScanMountGoat** [reply]
That still doesn't answer my question. What kind of effect are you trying to do? Is this still the cel shading shader? If you need to "jump through hoops", it could just be a simple misunderstanding that's easy to fix.

> The main screen texture isn't passed to the shader code in renderdoc

Forward rendering works well with alpha blending as long as you get the ordering right, so there's usually no need to pass in the current rendered image to each shader for layering or blending.

> If it means anything though, in the nufxlb, there is a "final" render pass, listed before the rest of the near far opaque etc variants, and this applies to each material.

Each model is only drawn once or twice if you count the depth only pass for shadow mapping. The tag at the end like "_opaque" or "_sort" just affects the order of the draw calls. We call these "passes", but not all of them result in a different render pass in RenderDoc. The "_far" and "_sort" draw calls share a render pass, for example.

> Anyways, as for the blending itself, that isn't done in the code for each material as far as I can tell. 

Alpha blending is controlled by the material. This is the blending state in the fragment buffer tab of the pipeline state in RenderDoc. 

The "layering" with different render passes is just done by taking in one or more input textures, modifying the values, and writing to an output texture. A "render pass" in a graphics API mostly just defines the input textures called "attachments" and the output textures called "targets". This is how post processing works in game. Draw calls that have the same input and output textures usually use the same render pass.

Some effects just draw the input texture to a quad covering the screen and use alpha blending to "blend" with the existing screen texture. I believe this is used for some of the bloom shaders, but it's been a while since I've looked at it.

**Reply by Trainmaster9977** [reply]
Ok, let me be more clear on what I'm trying to do in one message;

I am doing the cell shader. It needs to be rendered via Near to avoid the bloom pass, but needs to have the effects layered over it. Therefore, I am trying to find a way to rewrite the shader pipeline to save the effects on screen as a texture that I can then overlay onto specific meshes, so I can overlay it on top of the fighter within the code itself, thus avoiding any issues with layering. 

This seems pretty out there, but hey, so is about everything else I've done so far.

**Reply by ScanMountGoat** [reply]
You don't need to use the near pass to avoid bloom. You can just set the colors dark enough to avoid bloom. There's no need to rewrite the game's rendering passes. You can check on the material research website for the threshold value (about 70-75%). You can either have users set the texture or material colors dark enough, or darken the colors in the shader by multiplying by a value between 0.0 and 1.0. You could place the appropriate instructions right after accessing the col map color or at the very end of the shader before the final color value is written.

**Reply by ScanMountGoat** [reply]
I've also discussed this previously on https://github.com/ScanMountGoat/Smush-Material-Research/discussions/40#discussioncomment-5809971

**Reply by Trainmaster9977** [reply]
The problem with that is that the colors will then be inaccurate, though.

Anyways, sorry for making this thread spiral out of control, just wanted to ask about the 3D_final

**Reply by ScanMountGoat** [reply]
I would reference the reply I linked. This seems to be one of the harder concepts to grasp with more physically based rendering, causing a lot of confusion in the community. 

What do you mean by "accurate" colors? The final color on screen will depend on the lighting intensity for each stage unless you're using something like emissive textures. The col map color is just the percentage of this light used for diffuse shading or specular for metals. If you want a color of 255,255,255 on screen for most stages, you don't need the texture color to be as bright as 255,255,255. A slightly darker color will still render as white. 

The bloom threshold and the colors that render as white conveniently match up most of the time at roughly 75% brightness. Most of the blooming is from colors that would render as brighter than full white. You typically need to darken colors for them to be "accurate" in game unless they come from an existing PBR engine or program. You can hard clip the brightness in your shader, but that won't necessarily result in "accurate" colors. I would try different darkening amounts until you find one that works well or just ask users to experiment with the material parameters that control color gain.
https://scanmountgoat.github.io/Smush-Material-Research/materials/model-brightness/

**Reply by Trainmaster9977** [reply]
ah. Didn't know that somehow....that is totally my bad

Wow I keep rewriting my message, ok so tested a few things out.

First, I wanted to ask some questions- specifically whether that reflection was controlled by anything within the shader itself, as in my documentation there was stuff like this:

```
    temp_635 = fma(ColorOutput_R, temp_536, fp_c1_data[3].w);
    temp_638 = fma(ColorOutput_G, temp_620, fp_c1_data[3].w);
    temp_640 = fma(ColorOutput_B, temp_636, fp_c1_data[3].w);

    temp_639 = fma(temp_635, charColorGrading.y, 0.0900000036);
    temp_641 = fma(temp_638, charColorGrading.y, 0.0900000036);
    temp_643 = fma(temp_640, charColorGrading.y, 0.0900000036);

    temp_646 = (temp_639 * 0.212599993) + (temp_641 * fp_c1_data[3].x) + (temp_643 * fp_c1_data[3].y);
```

Which I wondered about whether it had anything to do with this, especially because I cut that step out in my rewritten shader, and asked if you had any guesses.

But then, I was like, _Ok, I'll just test this out myself really quick_

So second off, I tested out trying to just pass a random specific color to the output attribute of the shader in renderdoc. And sure enough, like you said, it wasn't accurate. 

I checked both the pre-lut(?) and post-lut(?) passes, and it seemed to be inaccurate for both. (The lut seemed to be the point where stuff diverged)

So I started being like _Holy hell you're right_ and was typing up a message reflecting that, saying that I got it all wrong and where to look to rewrite the shaders that might control that inaccuracy/reflection.

And then while typing, I realized one more thing. 

In the pre-lut(??) part, there was a button that said "Override display of linear data in gamma space". I often just test stuff in this mode anyways but had it turned off for whatever reason. 

So third off, when I turned it back on, and tested the color accuracy of passing colors directly to an output attribute......

It was 1:1.

Exactly.

And then of course it diverges post LUT, and when converting to I assume SRGB properly?

Either way, there is a point in it where you can in fact just give it an input and it'll render correctly....but then the shenanigans happen around the time that the LUT does.

So, I guess I want to ask some questions about how this actually works color wise. Sorry for my ignorance on this part of shaders.

**Reply by ScanMountGoat** [reply]
All the testing should be with the input texture colors. Writing to fragment output will just give you the exact color you wrote when you view the output texture for that pass in the texture view. There's not really a "correct" way to visualize a texture, so RenderDoc gives a lot of options for gamma, brightness, channels, etc. It should show the actual value stored in the image data if you click a spot on the texture view. 

> specifically whether that reflection was controlled by anything within the shader itself

Even if you output a texture color directly to the output color attribute, you'll still have color adjustments from post processing passes and bloom to deal with. The colors appearing brighter than intended isn't really a shader specific issue. It's a fundamental part of the game's rendering. You can change it but not in a way that only affects certain shaders without a lot of work. 

One of the benefits of more physically based renderers is that materials look consistent across different lighting conditions. If you use a plausible value for the albedo or base color, you should get reasonable results across stages. Getting an "accurate" final color depends on what lighting you use as a reference.

**Reply by Trainmaster9977** [reply]
> You can change it but not in a way that only affects certain shaders without a lot of work.

So you're saying there's a chance-

To be serious though, I would like to ask in that case if you have any places for me to look for that besides renderdoc itself (which I will take a look at obviously, always good to have the actual files though), or any theoretical methods you'd suggest.

**Reply by ScanMountGoat** [reply]
If you have a case where the model with the cel shading shader is blooming, try just making the texture colors darker first. You can make a capture in RenderDoc with that specific instance of unwanted blooming in Ryujinx and edit the shaders in RenderDoc to see how you need to adjust them. Trying out different values for the col map should also help build more intuition about the interaction between the input and output colors and how it relates to bloom.

Keep in mind that the texture values may work differently than you expect if the texture uses an sRGB format. An sRGB texture value of 0.5 would be accessed in the shaders as roughly 0.5^2.2 (you can find the exact sRGB to linear conversion online). This is how the comparison images were made for the model brightness tutorial page. Comparing the texture color vs final color on screen for several different values may also be helpful for learning purposes.

It sounds like your goal is just to make the shader have the intended colors and avoid unwanted blooming. I'm providing a simple suggestion to try to help you work through these issues. If the simple method doesn't work, then I can try and provide more guidance on a more involved fix. I don't want you to waste time reworking large portions of the game's rendering if possible. If you're interested in learning about it for other reasons, feel free to let me know what progress you make or if you have any questions.

**Reply by Trainmaster9977** [reply]
Well yeah, that's my goal, but I also don't want any workarounds that only *usually* work.

Speaking of, however, would changing the textures to use linear solve the SRGB issue? Additionally, how does that play with the fact that smash uses SRGB as a whole?
